### ISSUE TITLE
Disable menu items when action is not possible

### ISSUE THREAD

**Issue by pranavmodx** [4ian/GDevelop#1550]
## Describe the bug
Not exactly a bug but a possible enhancement/prevention of future bugs. The corresponding menu item should be disabled when an action is not possible in a given setting. Else, it could become a reason for a bug at some other time when user clicks the menu item when it isn't allowed/possible to.

## To Reproduce
1. Open desktop app
2. Click on Edit Menu or any other menu
3. Observe all menu items are by default enabled

## Screenshots
![Screenshot 2020-03-18 at 9 59 47 PM](https://user-images.githubusercontent.com/42793024/76984730-4052c100-6965-11ea-9ba8-8acd715f4ee7.png)
Eg: Here, actions like cut, copy, paste should not possible (as this is the start page).

![Screenshot 2020-03-18 at 10 00 29 PM](https://user-images.githubusercontent.com/42793024/76984860-6aa47e80-6965-11ea-95d0-b36e1092a009.png)
Example of an app (MS Todo) which disables menu items accordingly.

## Other details
* Mac OSX Mojave
* Desktop App

**Comment by 4ian** [comment]
For this kind of issues to be useful, it would be a good idea to list what to do, so that the issue is actionable - otherwise this issue will surely stay there forever.
Opening an issue is good, but making the work of finding what is really missing in details is even better :)

**Comment by pranavmodx** [comment]
> For this kind of issues to be useful, it would be a good idea to list what to do, so that the issue is actionable - otherwise this issue will surely stay there forever.

Thanks for mentioning this. Will keep this in mind in the future. :)

Is there an easy workaround for this? Or does this require a lot of changes with conditionals for different scenarios?

**Comment by 4ian** [comment]
Probably the work is quite different for each menu item, might not be totally obvious how to get the info to know if we can disable the item or not. Look at ElectronMainMenu.js to get an idea :)

**Comment by Minisonny** [comment]
Is it only me or anyone else also observe that the _Edit_ dropdown doesn't work even in the main editor 😕

**Comment by 4ian** [comment]
It's only for text fields.

**Comment by 4ian** [comment]
But to be fair yeah it's confusing because it's always enabled even when not useful!

**Comment by pranavmodx** [comment]
The File Menu and the Edit Menu are the most confusing ones I think. So maybe fixing just those at this time would help?

I'm not sure regarding the solution but here is what I worked out :

- When to disable to each menu?
1. File Menu [items such as save, save as, export, close]  - At the start page and/or when no project is open.
2. Edit Menu [probably all items] - At the start page and in the main editor when object instance is selected (as you mentioned that the edit menu is only for text fields).

Note : There may be many other cases where you might want to disable the menu(s), but that is if the solution is valid.

- How to disable the menus?
-> There is [remote](https://www.electronjs.org/docs/api/remote) api by which we can dynamically change the attributes ('enabled' in this case) of menu/submenu inside a web page.
-> This [article](https://www.electronjs.org/docs/api/menu#instance-properties) shows an example of it.

@4ian Please let me know your thoughts on this.

**Comment by Minisonny** [comment]
Yeah, the `enabled` is used various places with the `label` and self-defined click function. But in `ElectronMainMenu` it is predefined with `role`. I'm not sure but maybe we can try to check the `Clipboard` to see if we should enable or not

**Comment by 4ian** [comment]
**Here is how it works right now:**

* The app is controlling the menu with ElectronMainMenu.js
* It's sending the menu to electron with an event, and then electron (see MainMenu.js) renders the menu (this allows to separate the electron only code in `electron-app`).
* When there is a click, an event is sent back from MainMenu.js to ElectronMainMenu.js, which calls editor methods.

It works ok, but alas the way I did this originally is making the menu not very "reactive". See `LocalApp.js`. The menu is containing MainFrame. This means that the menu does not know anything about the currently opened project, etc... It can only call some methods of MainFrame. Not very flexible!

**If you want to have the main menu more reactive, you probably want to:**

1. refactor ElectronMainMenu and MainFrame so that ElectronMainMenu is not "containing" MainFrame, but the reverse. You pass ElectronMainMenu as a prop to MainFrame, like something named `renderMainMenu`. The reason? So that MainFrame can pass some props to ElectronMainMenu, like the current project. You have also to pass additional props for all `this._editor.xxx`
2. using these props, you can call _buildAndSendMenuTemplate (it's called in componentDidUpdate) and add `enabled` according to the props that are passed to ElectronMainMenu.
3. That's it :D The refactoring is the longest part I think :) Be sure to call `_buildAndSendMenuTemplate` only when really needed, to avoid spending time rebuilding the menu if nothing changed.

**Bonus**: by doing this, we can in the future make the web-app have a menu :D We would "just" have to copy ElectronMainMenu and replace it by a list of buttons/menus made with material-ui. Then, in BrowserApp.js, pass the material-ui powered MainMenu to MainFrame.

You can see that these changes will remove `React.cloneElement` in ElectronMainMenu. This is a sign that the new approach is much better in terms of components architecture.

**Comment by 4ian** [comment]
What I described works well for, I think, disabling menu items in File/Edit/etc... menus.
It might require more work to get the Edit menu to be dynamically updated according to the selection. I would say not to worry about this - starting by disabling the menu options when there is no project opened would be a good start :)

**Comment by 85ayush** [comment]
I would love to work on this issue, I am new to open source contribution, can you please guide me?
